core/src/main/java/org/elasticsearch/search/sort/ScriptSortBuilder.java

```diff
@@ -78,22 +146,189 @@ public ScriptSortBuilder setNestedPath(String nestedPath) {
+        builder.startObject(NAME);
+        builder.field(SCRIPT_FIELD.getPreferredName(), script);
+        builder.field(TYPE_FIELD.getPreferredName(), type);
         builder.field(ORDER_FIELD.getPreferredName(), order);
         if (sortMode != null) {
-            builder.field("mode", sortMode);
+            builder.field(SORTMODE_FIELD.getPreferredName(), sortMode);
         }
         if (nestedPath != null) {
-            builder.field("nested_path", nestedPath);
+            builder.field(NESTED_PATH_FIELD.getPreferredName(), nestedPath);
         }
         if (nestedFilter != null) {
-            builder.field("nested_filter", nestedFilter, builderParams);
+            builder.field(NESTED_FILTER_FIELD.getPreferredName(), nestedFilter, builderParams);
         }
         builder.endObject();
         return builder;
     }
+
+    @Override
+    public ScriptSortBuilder fromXContent(QueryParseContext context, String elementName) throws IOException {
+        ScriptParameterParser scriptParameterParser = new ScriptParameterParser();
+        XContentParser parser = context.parser();
+        ParseFieldMatcher parseField = context.parseFieldMatcher();
+        Script script = null;
+        ScriptSortType type = null;
+        String sortMode = null;
+        SortOrder order = null;
+        QueryBuilder<?> nestedFilter = null;
+        String nestedPath = null;
+        Map<String, Object> params = new HashMap<>();
+
+        XContentParser.Token token;
+        String currentName = parser.currentName();
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentName = parser.currentName();
+            } else if (token == XContentParser.Token.START_OBJECT) {
+                if (parseField.match(currentName, ScriptField.SCRIPT)) {
+                    script = Script.parse(parser, parseField);
+                } else if (parseField.match(currentName, PARAMS_FIELD)) {
+                    params = parser.map();
+                } else if (parseField.match(currentName, NESTED_FILTER_FIELD)) {
+                    nestedFilter = context.parseInnerQueryBuilder();
+                }
+            } else if (token.isValue()) {
+                if (parseField.match(currentName, ORDER_FIELD)) {
+                    order = SortOrder.fromString(parser.text());
+                } else if (scriptParameterParser.token(currentName, token, parser, parseField)) {
+                    // Do Nothing (handled by ScriptParameterParser
+                } else if (parseField.match(currentName, TYPE_FIELD)) {
+                    type = ScriptSortType.fromString(parser.text());
+                } else if (parseField.match(currentName, SORTMODE_FIELD)) {
+                    sortMode = parser.text();
+                } else if (parseField.match(currentName, NESTED_PATH_FIELD)) {
+                    nestedPath = parser.text();
+                }
```

Comment:
Exception for fall through? (Thought about it after reading the comment in line 205)